### PR TITLE
Better default worksheet size

### PR DESCRIPTION
**Pull request by jasonqng**
This PR gives users the option to adjust the dimensions of worksheets to fit the dataframe.

You can now optionally pass in the param `df_size = True` to the upload() function in order to have it adjust the sheet to the size of your dataframe. This can affect both new sheets and existing sheets.

Currently, the default with a new sheet is to create one that is 1000 rows by 100 columns. This is an issue because you would by default only be able to create 20 worksheets since workbooks are capped at 2 million cells, and you would have to manually delete potentially unnecessary rows and columns in order to create more than 20 worksheets.

With this PR, if you try to upload the below, you will create a worksheet that is exactly 3 rows by 5 columns.:
```
df = [pd.Series([1., 2., 3.], index=['a', 'b', 'c']),
    pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd'])]
d2g.upload(df, spreadsheet, wks_name, df_size = True)
```

Similarly, if you pass in `df_size = True` for an existing sheet, the sheet will automatically resize to match the dataframe's dimensions. This is possible using gspread's `resize()` function. Currently, `upload()` only allows for adding more rows and columns if the dataframe is bigger than the existing worksheet. This PR now allows for resizing the sheet smaller if the dataframe doesn't need all the space. This allows you to maximize the number of sheets you can create. 

If you do not pass `df_size` or pass `df_size = False`, the sheet size will not change unless the dataframe is larger, maintaining existing behavior. Thus, these changes will not affect any pre-existing jobs that use df2gspread since you must explicitly pass `df_size = True` to get the new behavior.

If you don't want to resize the sheet down to match the df's size but also want something different than the 1000x100 default, you can now also pass `new_sheet_dimensions = (num_rows, num_cols)` to `upload()` and this will override the default setting.

df_size should work with the pre-existing start_cell feature added in this PR (https://github.com/maybelinot/df2gspread/pull/17), but there may be possible edge cases where things break. I've added testing in this PR to check for continued proper behavior when these two are used together.

So to recap, for df_size: 
- If True and worksheet name does NOT exist, will create a new worksheet that is the size of the df.
- If False and the worksheet name does NOT exist, will create a default sheet of size 1000 x 100. If the dataframe ends up being larger, it will then resize only the dimension(s) that are larger up to fit the dataframe. It will not reduce pre-existing columns or rows.
- If True and worksheet does exist, will resize larger or smaller to fit new dataframe. 
- If False and dataframe is larger than existing sheet, will resize the sheet larger only on the dimension(s) that are larger. It will not reduce pre-existing columns or rows.
- If False and dataframe is smaller than existing sheet, does not resize.

### CODE REVIEW

**Comment by jasonqng** [comment]
To test the new functionality:
Checkout branch and from the df2gspread/ folder (the one with /bin, /docs, etc), in python shell:

```
import numpy as np
import pandas as pd
from df2gspread import df2gspread as d2g
spreadsheet = 'YOUR_SPREADSHEET'
small_df = pd.DataFrame(data = np.array([np.arange(10)]*1).T)
big_df = pd.DataFrame(data = np.array([np.arange(2000)]*1).T)

######## TESTING BEHAVIOR OF INSERTING INTO NEW WORKSHEETS #########

# 1a: Insert into first worksheet without df_size, df is smaller than default size
# Matches existing behavior of creating 1000x100 sheet and filling only small portion of it.
d2g.upload(small_df, spreadsheet)

# 1b: Insert into first worksheet without df_size, df is larger than default size
# Sheet expands to accommodate all 2000 rows but it still has 100 columns, matching pre-existing behavior
d2g.upload(big_df, spreadsheet)

# 2a: Insert into new worksheet without df_size, df is smaller than default size
# Matches existing behavior of creating 1000x100 sheet and filling only small portion of it.
wks_name = 'Test2a'
d2g.upload(small_df, spreadsheet, wks_name)

# 2b: Insert into new worksheet without df_size, df is larger than default size
# Sheet expands to accommodate all 2000 rows but still has 100 columns, matching pre-existing behavior
wks_name = 'Test2b'
d2g.upload(big_df, spreadsheet, wks_name)

# 3a: Insert into new worksheet WITH df_size, df is smaller than default size
# Creates a sheet that has the same dimension as the df 
wks_name = 'Test3a'
d2g.upload(small_df, spreadsheet, wks_name, df_size = True)

# 3b: Insert into new worksheet WITH df_size, df is larger than default size
# Creates a sheet that has the same dimension as the df 
wks_name = 'Test3b'
df = pd.DataFrame(data = np.array([np.arange(2000)]*1).T)
d2g.upload(df, spreadsheet, wks_name, df_size = True)

######## TESTING BEHAVIOR OF INSERTING INTO EXISTING WORKSHEETS #########

# 4a: Insert into existing worksheet without df_size, df is smaller
# No change in size
wks_name = 'Test2a'
d2g.upload(small_df, spreadsheet, wks_name)

# 4b: Insert into existing worksheet without df_size, df is larger
# Sheet expands to accommodate all 2000 rows retains number of columns in original, matching pre-existing behavior
wks_name = 'Test2a'
d2g.upload(big_df, spreadsheet, wks_name)

# 5a: Insert into existing worksheet WITH df_size, df is smaller
# Sheet is sized down to match the df
wks_name = 'Test2a'
d2g.upload(small_df, spreadsheet, wks_name, df_size = True)

# 5b: Insert into existing worksheet WITH df_size, df is larger
# Sheet expands to match the df
wks_name = 'Test2a'
d2g.upload(big_df, spreadsheet, wks_name, df_size = True)

**Comment by jasonqng** [comment]
Tests:

```
➜  df2gspread git:(master) ✗ git checkout jasonqng:better-default-worksheet-size
error: pathspec 'jasonqng:better-default-worksheet-size' did not match any file(s) known to git.
➜  df2gspread git:(master) ✗ git checkout better-default-worksheet-size
Switched to branch 'better-default-worksheet-size'
➜  df2gspread git:(better-default-worksheet-size) ✗ py.test tests/ -v
========================================== test session starts ==========================================
platform darwin -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /Users/jpersonal/Projects/df2gspread, inifile:
collected 8 items

tests/test_credentials.py::test_client_secrets_credentials xfail
tests/test_df2gspread.py::test_global_import PASSED
tests/test_df2gspread.py::test_version_check PASSED
tests/test_df2gspread.py::test_spreadsheet_invalid_file_id PASSED
tests/test_df2gspread.py::test_worksheet_invalid_name PASSED
tests/test_df2gspread.py::test_worksheet PASSED
tests/test_df2gspread.py::test_big_worksheet PASSED
tests/test_df2gspread.py::test_delete_file PASSED

================================= 7 passed, 1 xfailed in 86.16 seconds ==================================
```

File: df2gspread/df2gspread.py [hunk 1]
```diff
@@ -102,12 +102,18 @@ def upload(df, gfile="/New Spreadsheet", wks_name=None, chunk_size=1000,
...
-    if df_size or len(df.index) + 1 > wks.row_count or len(df.columns) + 1 > wks.col_count:
+    # If user requested to resize sheet to fit dataframe, go ahead and 
+    # resize larger or smaller to better match new size of pandas dataframe.
+    # Otherwise, leave it the same size unless the sheet needs to be expanded
+    # to accomodate a larger dataframe.
+    if len(df.index) + 1 > wks.row_count:
+        wks.add_rows(len(df.index) - wks.row_count + 1)
+    if len(df.columns) + 1 > wks.col_count:
+        wks.add_cols(len(df.columns) - wks.col_count + 1)
+    if df_size:
```
Comment (maybelinot) [inline df2gspread/df2gspread.py]: Does it make sense to move `if df_size` before other `if` conditions?

Comment (jasonqng) [inline df2gspread/df2gspread.py]: Yes, makes sense. I think I put it after because I was dealing with a bug where if the sheet was full (same amount of data as cells), the clean function would add extra rows and cols and I compensated by resizing after. But I think I fixed the bug in this latest commit by only adding extra rows and cols if needed.

Comment (maybelinot) [inline df2gspread/df2gspread.py]: Okay, I need to review it. And it wouldn't be bad to write some test cases to test new features (start_cell and df_size).

File: df2gspread/df2gspread.py [hunk 2]
```diff
@@ -44,6 +55,8 @@ def upload(df, gfile="/New Spreadsheet", wks_name=None, chunk_size=1000,
         :type row_names: bool
         :type clean: bool
         :type credentials: class 'oauth2client.client.OAuth2Credentials'
+        :type df_size: bool
+        :type df_size: tuple
```
Comment (maybelinot) [inline df2gspread/df2gspread.py]: Did you mean `new_sheet_dimension`?

File: df2gspread/df2gspread.py [hunk 3]
```diff
@@ -69,7 +82,16 @@ def upload(df, gfile="/New Spreadsheet", wks_name=None, chunk_size=1000,
...
         gfile_id = get_file_id(credentials, gfile, write_access=True)
 
-    wks = get_worksheet(gc, gfile_id, wks_name, write_access=True)
+    # Tuple of rows, cols in the dataframe.
+    # If user did not explicitly specify to resize sheet to dataframe size
+    # then for new sheets set it to new_sheet_dimensions, which is by default 1000x100
+    if df_size:
+        new_sheet_dimensions = (len(df), len(df.columns))
+    else:
+        new_sheet_dimensions = new_sheet_dimensions
```
Comment (maybelinot) [inline df2gspread/df2gspread.py]: What the purpose of this line?

File: df2gspread/df2gspread.py [hunk 4]
```diff
@@ -82,12 +104,18 @@ def upload(df, gfile="/New Spreadsheet", wks_name=None, chunk_size=1000,
...
-    if len(df.columns) + 1 > wks.col_count:
-        wks.add_cols(len(df.columns) - wks.col_count + 1)
+    # If user requested to resize sheet to fit dataframe, go ahead and 
+    # resize larger or smaller to better match new size of pandas dataframe.
+    # Otherwise, leave it the same size unless the sheet needs to be expanded
+    # to accomodate a larger dataframe.
+    extra_col = 1 if col_names else 0
+    extra_row = 1 if row_names else 0
+    if df_size:
+        wks.resize(rows=len(df.index) + extra_row, cols=len(df.columns) + extra_col)
```
Comment (maybelinot) [inline df2gspread/df2gspread.py]: `len(df.index) + row_names` gives the same result as `len(df.index) + extra_row`..
I'm not sure that `extra_row = 1 if row_names else 0` adds more sense.. Would it be better to get rid of these two lines?

File: df2gspread/gfiles.py [hunk 5]
```diff
@@ -77,7 +77,8 @@ def get_worksheet(gc, gfile_id, wks_name, write_access=False):
             wks = spsh.worksheet(wks_name)
         else:
             if write_access == True:
-                wks = spsh.add_worksheet(wks_name, 1000, 100)
+                rows, cols = new_sheet_dimensions
+                wks = spsh.add_worksheet(wks_name, rows, cols)
```
Comment (maybelinot) [inline df2gspread/gfiles.py]: Also could be improved to `spsh.add_worksheet(wks_name, *new_sheet_dimensions)`
And checking of size of new_sheet_dimensions wouldn't be bad

**Comment by maybelinot** [comment]
BTW, @jasonqng can you squash your commits after updating of this PR? One commit for feature plus one for testing will be enough..

File: df2gspread/df2gspread.py (hunk 3 above)
Comment (jasonqng) [inline df2gspread/df2gspread.py]: Good catch. Leftover, will remove.

File: df2gspread/df2gspread.py (hunk 4 above)
Comment (jasonqng) [inline df2gspread/df2gspread.py]: Much cleaner, agreed.

File: df2gspread/gfiles.py (hunk 5 above)
Comment (jasonqng) [inline df2gspread/gfiles.py]: Not sure what you mean here. `*new_sheet_dimensions`? And what kind of checking? Do you mean checking to make sure someone doesn't put in a value too large?

Comment (maybelinot) [inline df2gspread/gfiles.py]: This part 
```
rows, cols = new_sheet_dimensions
wks = spsh.add_worksheet(wks_name, rows, cols)
```
can be replased with this line
`spsh.add_worksheet(wks_name, *new_sheet_dimensions)`

Yes, checking of size of new_sheet_dimensions..

Comment (jasonqng) [inline df2gspread/gfiles.py]: Oh durf, right, args. Sure, I'll add basic checking but I assume it's perfectly acceptable to let gspread or Google handle the actual reporting of the error, if there is one, just like in the rest of the package.

Comment (jasonqng) [inline df2gspread/gfiles.py]: I think the error message by Google below are pretty clear. I guess you could wrap it in a try...except and retry it with default dimensions if the user input returns one of these errors, but I think that just obfuscates the issue. Better to let the user know an issue occurred. But feel free to add in error checking if you think it's necessary, happy to defer to you!

```
>>> d2g.upload(df, spreadsheet, wks_name, credentials=credentials, clean=False, new_sheet_dimensions=(10000000,100000000))
...
gspread.exceptions.RequestError: 400: This action would increase the number of columns in the sheet above the limit of 18278 columns.
```
```
>>> d2g.upload(df, spreadsheet, wks_name, credentials=credentials, clean=False, new_sheet_dimensions=('a','b'))
...
gspread.exceptions.RequestError: 400: [Line 1, Column 194, element gs:rowCount] The count must be specified.
```
Full traceback
```
>>> d2g.upload(df, spreadsheet, wks_name, credentials=credentials, clean=False, new_sheet_dimensions=(0,0))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "df2gspread/df2gspread.py", line 95, in upload
    new_sheet_dimensions=new_sheet_dimensions)
  File "df2gspread/gfiles.py", line 81, in get_worksheet
    wks = spsh.add_worksheet(wks_name, *new_sheet_dimensions)
  File "/Library/Python/2.7/site-packages/gspread/models.py", line 89, in add_worksheet
    elem = self.client.post_feed(url, ElementTree.tostring(feed))
  File "/Library/Python/2.7/site-packages/gspread/client.py", line 228, in post_feed
    raise RequestError(ex.message)
gspread.exceptions.RequestError: 400: [Line 1, Column 194, element gs:rowCount] The count must be specified.

**Comment by jasonqng** [comment]
@maybelinot Feature commits squashed (except for the latest to show you recent changes) and testing added. Let me know if you need anything else. Thanks for the speedy and excellent code review! Can't wait to get this feature added.

File: tests/test_df2gspread.py [hunk 6]
```diff
@@ -310,6 +310,131 @@ def test_df2gspread_start_cell(user_credentials_not_available):
...
+        random.choice(string.ascii_uppercase + string.digits)
+        for _ in range(10))
+    credentials = get_credentials()
+    gc = gspread.authorize(credentials)
+    gfile_id = get_file_id(credentials, filepath, write_access=True)
+
+    df_upload_a = pd.DataFrame(
+        {0: ['1', '2', 'x', '4']},
+        index=[0, 1, 2, 3])
+    df_upload_a = df_upload_a.fillna('0')
```
Comment (maybelinot) [inline tests/test_df2gspread.py]: Why are you using `fillna` method here?

Comment (jasonqng) [inline tests/test_df2gspread.py]: Good point. Just copied and pasted other test. You're right, doesn't do anything, removed and test still passes.

Comment (maybelinot) [inline tests/test_df2gspread.py]: I added this method to fill empty dataframe with zeros.. for example [here](https://github.com/jasonqng/df2gspread/blob/f8e5e85388b990389f037d13386c96e22d4fa809/tests/test_df2gspread.py#L209). In case of filled df it do nothing

**Comment by maybelinot** [comment]
@jasonqng, nice work man!
Thanks a lot for contribution to this project!

**Comment by jasonqng** [comment]
Glad to be of help!
